### PR TITLE
Degree should have a "non-official" or "external" Code

### PR DESCRIPTION
**Pull request by NunoPinheiro**
Closes [ACADEMIC-626](https://jira.fenixedu.org/browse/ACADEMIC-626)

Takes into account suggestions given in #637 except for the "Optional" one

### CODE REVIEW

**Comment by jcarvalho** [comment]
Please squash all your commits into one (you don't need to redo the pull request, just amend your branch).

**Comment by pedrosan7os** [comment]
I'm also making a change in the name and details of the PR for easier handling, if you could follow in the next ones would be awesome.

File: src/main/java/org/fenixedu/academic/service/services/bolonhaManager/CreateCompetenceCourse.java [hunk 1]
```diff
@@ -42,6 +42,26 @@
 
     @Atomic
     public static CompetenceCourse run(String name, String nameEn, String acronym, Boolean basic, RegimeType regimeType,
+            CompetenceCourseLevel competenceCourseLevel, CompetenceCourseType type, String unitID,
```
Comment (jcarvalho) [inline src/main/java/org/fenixedu/academic/service/services/bolonhaManager/CreateCompetenceCourse.java]: With this change, the code in the method will be duplicated with the newly deprecated one, minus the `setCode`. You should reuse some of the code.

File: src/main/java/org/fenixedu/academic/service/services/bolonhaManager/EditCompetenceCourse.java [hunk 2]
```diff
@@ -52,6 +53,15 @@ protected void run(String competenceCourseID, String name, String nameEn, Boolea
         competenceCourse.edit(name, nameEn, basic, competenceCourseLevel, type, curricularStage);
     }
 
+    protected void run(String competenceCourseID, String name, String nameEn, Boolean basic,
+            CompetenceCourseLevel competenceCourseLevel, CompetenceCourseType type, CurricularStage curricularStage, String code)
+            throws FenixServiceException {
+        final CompetenceCourse competenceCourse = readCompetenceCourse(competenceCourseID);
```
Comment (jcarvalho) [inline src/main/java/org/fenixedu/academic/service/services/bolonhaManager/EditCompetenceCourse.java]: Same as [above](https://github.com/FenixEdu/fenixedu-academic/pull/645/files#diff-4ce2e28b012b63023627a81d6c4016c2R45).

File: src/main/java/org/fenixedu/academic/service/services/bolonhaManager/EditCompetenceCourse.java [hunk 3]
```diff
@@ -245,6 +255,26 @@ public static void runEditCompetenceCourse(String competenceCourseID, Integer ol
     @Atomic
     public static void runEditCompetenceCourse(String competenceCourseID, String name, String nameEn, Boolean basic,
             CompetenceCourseLevel enumCompetenceCourseLevel, CompetenceCourseType enumCompetenceCourseType,
+            CurricularStage valueOf, String code) throws FenixServiceException {
+        try {
```
Comment (jcarvalho) [inline src/main/java/org/fenixedu/academic/service/services/bolonhaManager/EditCompetenceCourse.java]: Same as [above](https://github.com/FenixEdu/fenixedu-academic/pull/645/files#diff-4ce2e28b012b63023627a81d6c4016c2R45).

File: src/main/resources/resources/ScientificCouncilResources_pt.properties [hunk 4]
```diff
@@ -347,6 +348,7 @@ label.message.invalid.line.format = A linha {0} foi ignorada pois n
 label.message.username.invalid = A linha {0} foi ignorada pois n�o contem um utilizador v�lido: {1}.
 label.message.lessonHours.invalid = A linha {0} foi ignorada pois n�o contem um valor num�rico de horas v�lido: {1}.
 label.message.teacherCategory.invalid = A linha {0} foi ignorada pois n�o contem uma categoria profissional v�lida: {1}.
+label.ministryCode=C�digo do Oficial (Minist�rio)
```
Comment (jcarvalho) [inline src/main/resources/resources/ScientificCouncilResources_pt.properties]: Should be only `Código Oficial`.

**Comment by jcarvalho** [comment]
The code could be greatly improved/simplified by the use of Java 8 and not using ancient APIs like Apache Commons, but otherwise everything looks fine.

**Pull request merged**